internal/xds/rbac/matchers.go

```diff
@@ -0,0 +1,407 @@
+	matcher2 "google.golang.org/grpc/internal/xds/matcher"
+)
+
+// TODO: Do we need errors at every node? Should we have checks on each node (i.e. destinationPort should be within a certain range)
+
+// The matcher interface will be used by the RBAC Engine to help determine
+// whether incoming RPC requests should be allowed or denied. There will be many
+// types of matchers, each instantiated with part of the policy used to
+// instantiate the RBAC Engine. These constructed matchers will form a logical
+// tree of matchers, which data about any incoming RPC's will be passed through
+// the tree to help make a decision about whether the RPC should be allowed or
+// not.
+type matcher interface {
+	matches(args *EvaluateArgs) bool
+}
+
+// policyMatcher helps determine whether an incoming RPC call matches a policy.
+// A policy is a logical role (e.g. Service Admin), which is comprised of
+// permissions and principals. A principal is an identity (or identities) for a
+// downstream subject which are assigned the policy (role), and a permission is
+// an action(s) that a principal(s) can take. A policy matches if both a
+// permission and a principal match, which will be determined by the child or
+// permissions and principal matchers.
+type policyMatcher struct {
+	permissions *orMatcher
+	principals  *orMatcher
+}
+
+func newPolicyMatcher(policy *v3rbacpb.Policy) (*policyMatcher, error) {
+	permissionsList, err := createMatcherListFromPermissionList(policy.Permissions)
+	if err != nil {
+		return nil, err
+	}
+	principalsList, err := createMatcherListFromPrincipalList(policy.Principals)
+	if err != nil {
+		return nil, err
+	}
+	return &policyMatcher{
+		permissions: &orMatcher{
+			matchers: permissionsList,
+		},
+		principals: &orMatcher{
+			matchers: principalsList,
+		},
+	}, nil
+}
+
+func (pm *policyMatcher) matches(args *EvaluateArgs) bool {
+	// Due to a policy matching iff one of the permissions match the action
+	// taking place and one of the principals match the peer, you can simply
+	// delegate the data about the incoming RPC to the child permission and
+	// principal or matchers.
+	return pm.permissions.matches(args) && pm.principals.matches(args)
+}
+
+// createMatcherListFromPermissionList takes a list of permissions (can also be
+// a single permission, e.g. from a not matcher which is logically !permission)
+// and returns a list of matchers which correspond to that permission. This will
+// be called in many instances throughout the initial construction of the RBAC
+// engine from the AND and OR matchers and also from the NOT matcher.
+func createMatcherListFromPermissionList(permissions []*v3rbacpb.Permission) ([]matcher, error) {
+	var matcherList []matcher
+	for _, permission := range permissions {
+		switch permission.GetRule().(type) {
+		case *v3rbacpb.Permission_AndRules:
+			andPermissionList, err := createMatcherListFromPermissionList(permission.GetAndRules().Rules)
```

Comment:
I would rename the local matcher used in all of these cases to `m`.

So, each of the cases would look as:
```golang
m, err := createMatcherFromSomething(...)
if err != nil {
  return nil, err
}
matchers := append(matchers, m)
```

It is very clear from the `createMatcherFromSomething` function as to what matcher is being created here. So `m` is a completely acceptable variable name here. And given its scope is <5 lines, all the more justified.

Same applies to the function below as well.

Reply:
Ah, great point :). Changed.

Reply:
Also renamed ......List to "mList."